### ISSUE TITLE
path_provider getDownloadsDirectory documentation doesn't say when it might return null

### ISSUE THREAD

**Issue by Hixie** [flutter/flutter#143238]
The `getDownloadsDirectory` function is defined as potentially returning null. The documentation talks about it potentially throwing an UnsupportedError if this is not supported on the current platform, and talks about the returned Directory maybe not existing, but doesn't say when it might return null. So it's not clear if this is a failure mode that one should handle, or what to do when it happens.

https://pub.dev/documentation/path_provider/latest/path_provider/getDownloadsDirectory.html

**Comment by stuartmorgan** [comment]
The intent is that `UnsupportedError` is for platforms that don't implement this at all (e.g., a platform that has no concept of a download directory), while `null` would mean that the implementation understands the concept, but that there's no such directory available.

That distinction is probably not worth handling two different ways, especially since the null case should be very rare, so we could consider changing all the currently nullable methods to be non-nullable, and convert nulls from the platform interface to `MissingPlatformDirectoryException` (and then documenting that).

**Comment by Hixie** [comment]
The docs also refer to the possibility of the path being valid but not existing, that's distinct from the case of the path not being available?

**Comment by stuartmorgan** [comment]
Yes; for instance on Linux you could configure your XDG settings to say your Download directory is `~/thisdirectoryhasneverexisted`, and then `path_provider` would return that valid but non-existent directory.

**Comment by Hixie** [comment]
What's an example of the path not being available? Something like Linux when no downloads directory has been configured at all?

**Comment by stuartmorgan** [comment]
Right now you could get it by running on a Linux without `xdg-user-dir`, or where that exists but fails when called.

In theory iOS/macOS could return an empty array from the search, but I would be pretty surprised if that's possible in practice.

**Comment by Hixie** [comment]
Cool. Those details would be great to include in the docs. Thanks for the explanations!

**Comment by familan** [comment]
In connection with getDownloadsDirectory, the current README on pub.dev shows Downloads being available on ALL platforms, but the code under the Examples tab tests explicitly for Android and iOS to report that getDownloadsDirectory is not available on these platforms. I don't know if it is really available for iOS, but the CHANGELOG shows it becoming available for Android in v2.1.4. If that is so, then the Example page needs to be updated.

**Comment by stuartmorgan-g** [comment]
@familan Please file a new issue; the example is separate from the Dart docs.

**Comment by theprantadutta** [comment]
Hi @stuartmorgan-g @Hixie — would it be useful if I took this on? The maintainer responses upthread already lay out exactly what to add. My plan:

1. Update the dartdoc on `getDownloadsDirectory` in `path_provider/lib/path_provider.dart` and the matching method on `PathProviderPlatform` in `path_provider_platform_interface` to explain:
   - `UnsupportedError` is thrown when the platform has no concept of a "downloads directory"
   - `null` is returned when the platform supports the concept but no such directory is available (concrete example: Linux without `xdg-user-dir` configured, or where it fails when called)
   - the returned `Directory` may not exist (the existing case, e.g. an XDG-configured directory pointing somewhere non-existent)
2. Skip the README/example tangent — that's tracked separately per @stuartmorgan-g's earlier note.

Happy to defer if someone's already on this. Otherwise I'll open a PR and link back here.
